### DISCUSSION TITLE
Implicit recursive type alias error diagnostic for unsupported syntax

### DISCUSSION THREAD

**Discussion by Molkree** [microsoft/pyright#4328] [Ideas]
Suppose I want to define a recursive type that is a list that contains zero or more comma-separated values (either integers or other lists) (it's from [Advent of Code](https://adventofcode.com/2022/day/13)).
Examples:
```py
[1,1,3,1,1]
[[1],[2,3,4]]
[[[]]]
[1,[2,[3,[4,[5,6,7]]]],8,9]
```
I tried to write it like this:
```py
T = list[int | "T"]
```
pyright determined the type to be `Type[list[int | ValueType]]`. However, this code results in a runtime error: `TypeError: unsupported operand type(s) for |: 'type' and 'str'`

![image](https://user-images.githubusercontent.com/12860284/207365061-3b91214e-6859-4889-8c00-dbd9f9a9a738.png)

It works if I replace new and fancy `|` syntax with `Union`, like this:
```py
T = list[Union[int, "T"]]
```

So my question is: Is it possible to get the warning that it's invalid syntax in this case?

I'm on Python 3.11.1 and using Pylance in VS Code, same results with the latest Release (v2022.12.20) and Preview (v2022.12.21) versions. I'm not entirely sure to which Pyright versions they correspond, probably 1.1.282 and 1.1.283 respectively.

P.S. On a side note, is it possible to somehow get the Pyright version in the code? Or determine the version that's bundled with Pylance? There were no mentions of Pyright versions anymore in the Pylance changelog :(

**Comment by erictraut** [comment]
Interesting, this looks to me like a bug in the Python runtime. PEP 604, which introduced the `|` syntax for unions, didn't discuss this limitation.

@JelleZijlstra, what are your thoughts on this? Under what circumstances can a forward-declared (stringified) type not be used with the `|` operator? Is it disallowed in all non-annotation contexts? Is this behavior specified anywhere that you know of? Or is this behavior unintended — and therefore a bug? Do you think that type checkers should detect and report this condition, since it will lead to a runtime `TypeError` exception?

**Reply by JelleZijlstra** [reply]
At runtime the `|` operator on types uses https://github.com/python/cpython/blob/985a71032bf055240e61259285a0b4925c925383/Objects/unionobject.c#L157, which only allows `None`, types, generic aliases (e.g. `list[int]`) and other `|`-created unions. Strings, of course, are none of those things. Objects defined in `typing.py`  are generally more permissive; their `__or__`/`__ror__` methods delegate to `typing.Union` which accepts almost anything. As a result, `typing.List[int] | "X"` works but `list[int] | "X"` doesn't.

I wasn't involved in implementing PEP 604 so I don't know exactly why it was implemented this way. PEP 604 says "The existing typing.Union and | syntax should be equivalent.", so it's reasonable to argue that the current implementation doesn't follow the PEP.

**Reply by Molkree** [reply]
Thanks for the replies, @erictraut @JelleZijlstra.

I've found [the corresponding issue](https://github.com/python/cpython/issues/90015) in CPython repo. From what I can gather there, the consensus was that runtime behavior won't be changed and instead docs (and PEP?) should be clarified to mention that this use of forward references with `|` is not supported.

However, no action has been taken yet.

**Reply by erictraut** [reply]
OK, if this is the intended behavior and is unlikely to be changed, then it makes sense for pyright to detect and report the issue to save users from an unexpected runtime exception. This will be included in the next release of pyright.
